File: src/nova_audio_agent/realtime/volcengine/tts.py
"""Streaming text preparation for Doubao bidirectional TTS."""

from __future__ import annotations

import asyncio
import json
import re
from collections.abc import AsyncIterator, Callable
from dataclasses import dataclass
from typing import Any
from uuid import uuid4

from websockets.asyncio.client import connect as websocket_connect

from nova_audio_agent.realtime.volcengine.protocol import EventType, MessageType, VolcMessage

_BOUNDARY = re.compile(r"[，。！？；：,.!?;:]|\n")
DEFAULT_CONNECT_TIMEOUT = 20.0
DEFAULT_RECEIVE_TIMEOUT = 15.0


class TextChunker:
    def __init__(self, *, soft_limit: int = 18, hard_limit: int = 48) -> None:
        if soft_limit < 1 or hard_limit < soft_limit:
            raise ValueError("invalid TTS text limits")
        self._soft_limit = soft_limit
        self._hard_limit = hard_limit
        self._pending = ""
        self._first = True

    def push(self, text: str) -> tuple[str, ...]:
        if type(text) is not str:
            raise TypeError("TTS text delta must be a string")
        self._pending += text
        chunks: list[str] = []
        while self._pending:
            boundary = self._flush_boundary()
            if boundary is None and len(self._pending) < self._hard_limit:
                break
            end = self._hard_limit if boundary is None else min(boundary, self._hard_limit)
            chunks.append(self._pending[:end])
            self._pending = self._pending[end:]
            self._first = False
        return tuple(chunks)

    def finish(self) -> tuple[str, ...]:
        if not self._pending:
            return ()
        pending, self._pending = self._pending, ""
        self._first = False
        return (pending,)

    def _flush_boundary(self) -> int | None:
        for match in _BOUNDARY.finditer(self._pending):
            end = match.end()
            if self._first or end >= self._soft_limit:
                return end
        return None


class DoubaoTtsError(RuntimeError):
    """Sanitized bidirectional TTS failure."""


@dataclass(frozen=True, slots=True)
class TtsAudio:
    pcm: bytes


def tts_headers(*, api_key: str, resource_id: str) -> dict[str, str]:
    if not api_key or not resource_id:
        raise ValueError("TTS credentials and resource id are required")
    return {
        "X-Api-Key": api_key,
        "X-Api-Resource-Id": resource_id,
        "X-Api-Connect-Id": str(uuid4()),
    }


class DoubaoTtsSession:
    def __init__(
        self,
        *,
        websocket: Any,
        session_id: str,
        voice: str,
        sample_rate: int,
        receive_timeout: float,
    ) -> None:
        self._websocket = websocket
        self._session_id = session_id
        self._voice = voice
        self._sample_rate = sample_rate
        self._receive_timeout = receive_timeout
        self._finished = False

    async def send_text(self, text: str) -> None:
        if self._finished:
            raise DoubaoTtsError("豆包 TTS 会话已经结束")
        if type(text) is not str or not text:
            raise ValueError("TTS text must be non-empty")
        payload = _tts_payload(
            self._voice,
            self._sample_rate,
            event=EventType.TASK_REQUEST,
            text=text,
        )
        await self._websocket.send(
            VolcMessage(
                MessageType.FULL_CLIENT_REQUEST,
                EventType.TASK_REQUEST,
                json.dumps(payload, ensure_ascii=False).encode(),
                self._session_id,
            ).marshal()
        )

    async def finish(self) -> None:
        if self._finished:
            return
        self._finished = True
        await self._websocket.send(
            VolcMessage(
                MessageType.FULL_CLIENT_REQUEST,
                EventType.FINISH_SESSION,
                b"{}",
                session_id=self._session_id,
            ).marshal()
        )

    async def cancel(self) -> None:
        if self._finished:
            return
        self._finished = True
        await self._websocket.send(
            VolcMessage(
                MessageType.FULL_CLIENT_REQUEST,
                EventType.CANCEL_SESSION,
                b"{}",
                session_id=self._session_id,
            ).marshal()
        )

    async def events(self) -> AsyncIterator[TtsAudio]:
        while True:
            try:
                raw = await asyncio.wait_for(self._websocket.recv(), timeout=self._receive_timeout)
            except Exception as exc:
                raise DoubaoTtsError(f"豆包 TTS 接收失败（{type(exc).__name__}）") from exc
            if type(raw) is not bytes:
                raise DoubaoTtsError("豆包 TTS 返回了无效帧类型")
            message = VolcMessage.unmarshal(raw)
            if message.session_id not in {None, self._session_id}:
                continue
            if message.message_type == MessageType.AUDIO_ONLY_SERVER and message.payload:
                yield TtsAudio(message.payload)
            elif message.event == EventType.SESSION_FINISHED:
                return
            elif message.message_type == MessageType.ERROR or message.event in {
                EventType.SESSION_FAILED,
                EventType.CONNECTION_FAILED,
            }:
                raise DoubaoTtsError("豆包 TTS 请求失败")

    async def close(self) -> None:
        try:
            try:
                await self._websocket.send(
                    VolcMessage(
                        MessageType.FULL_CLIENT_REQUEST,
                        EventType.FINISH_CONNECTION,
                        b"{}",
                    ).marshal()
                )
            finally:
                await self._websocket.close()
        except Exception as exc:
            raise DoubaoTtsError(f"豆包 TTS 关闭失败（{type(exc).__name__}）") from exc


class DoubaoTtsClient:
    def __init__(
        self,
        *,
        endpoint: str,
        api_key: str,
        resource_id: str,
        voice: str,
        output_sample_rate: int = 24_000,
        connect_timeout: float = DEFAULT_CONNECT_TIMEOUT,
        receive_timeout: float = DEFAULT_RECEIVE_TIMEOUT,
        connector: Callable[..., Any] = websocket_connect,
        id_factory: Callable[[], str] | None = None,
    ) -> None:
        if connect_timeout <= 0 or receive_timeout <= 0:
            raise ValueError("TTS timeouts must be positive")
        self._endpoint = endpoint
        self._api_key = api_key
        self._resource_id = resource_id
        self._voice = voice
        self._output_sample_rate = output_sample_rate
        self._connector = connector
        self._connect_timeout = float(connect_timeout)
        self._receive_timeout = float(receive_timeout)
        self._id_factory = (lambda: str(uuid4())) if id_factory is None else id_factory

    async def open(self) -> DoubaoTtsSession:
        websocket: Any | None = None
        try:
            websocket = await self._connector(
                self._endpoint,
                additional_headers=tts_headers(
                    api_key=self._api_key,
                    resource_id=self._resource_id,
                ),
                open_timeout=self._connect_timeout,
                close_timeout=1,
                max_size=10 * 1024 * 1024,
            )
            await websocket.send(
                VolcMessage(
                    MessageType.FULL_CLIENT_REQUEST,
                    EventType.START_CONNECTION,
                    b"{}",
                ).marshal()
            )
            await _expect_event(
                websocket,
                EventType.CONNECTION_STARTED,
                timeout=self._receive_timeout,
            )
            session_id = self._id_factory()
            await websocket.send(
                VolcMessage(
                    MessageType.FULL_CLIENT_REQUEST,
                    EventType.START_SESSION,
                    json.dumps(
                        _tts_payload(
                            self._voice,
                            self._output_sample_rate,
                            event=EventType.START_SESSION,
                        ),
                        ensure_ascii=False,
                    ).encode(),
                    session_id,
                ).marshal()
            )
            await _expect_event(
                websocket,
                EventType.SESSION_STARTED,
                session_id=session_id,
                timeout=self._receive_timeout,
            )
            return DoubaoTtsSession(
                websocket=websocket,
                session_id=session_id,
                voice=self._voice,
                sample_rate=self._output_sample_rate,
                receive_timeout=self._receive_timeout,
            )
        except BaseException as exc:
            if websocket is not None:
                try:
                    await websocket.close()
                except Exception:
                    pass
            if isinstance(exc, asyncio.CancelledError | DoubaoTtsError):
                raise
            if not isinstance(exc, Exception):
                raise
            raise DoubaoTtsError(f"豆包 TTS 建连失败（{type(exc).__name__}）") from exc


async def _expect_event(
    websocket: Any,
    event: EventType,
    *,
    session_id: str | None = None,
    timeout: float,
) -> None:
    raw = await asyncio.wait_for(websocket.recv(), timeout=timeout)
    if type(raw) is not bytes:
        raise DoubaoTtsError("豆包 TTS 握手返回了无效帧")
    message = VolcMessage.unmarshal(raw)
    if message.event != event or (session_id is not None and message.session_id != session_id):
        raise DoubaoTtsError("豆包 TTS 握手返回了意外事件")


def _tts_payload(
    voice: str,
    sample_rate: int,
    *,
    event: EventType,
    text: str | None = None,
) -> dict[str, Any]:
    payload: dict[str, Any] = {
        "event": event,
        "user": {"uid": str(uuid4())},
        "namespace": "BidirectionalTTS",
        "req_params": {
            "speaker": voice,
            "audio_params": {
                "format": "pcm",
                "sample_rate": sample_rate,
                "enable_timestamp": False,
            },
            "additions": json.dumps({"disable_markdown_filter": False}),
        },
    }
    if text is not None:
        payload["req_params"]["text"] = text
    return payload

File: runtime/src/realtime/volcengine/tts.ts
import { randomUUID } from 'node:crypto'
import { isWellFormed, stripLikePython } from '../../python-text.js'
import { MAX_REALTIME_TEXT } from '../protocol.js'
import { MAX_VOLCENGINE_WIRE_FRAME_BYTES, volcengineOutputPcm } from './audio.js'
import { EventType, MessageType, VolcMessage } from './protocol.js'
import {
  DEFAULT_VOLC_CLOSE_TIMEOUT_MS,
  DEFAULT_VOLC_CONNECT_TIMEOUT_MS,
  DEFAULT_VOLC_RECEIVE_TIMEOUT_MS,
  webSocketVolcBinaryConnector,
  type VolcBinaryConnector,
  type VolcBinarySocket,
} from './websocket.js'

const boundaries = new Set([...`，。！？；：,.!?;:\n`])

export class TextChunker {
  readonly #softLimit: number
  readonly #hardLimit: number
  #pending: string[] = []
  #first = true

  constructor(options: {readonly softLimit?: number; readonly hardLimit?: number} = {}) {
    const softLimit = options.softLimit ?? 18
    const hardLimit = options.hardLimit ?? 48
    if (!Number.isSafeInteger(softLimit) || !Number.isSafeInteger(hardLimit)
      || softLimit < 1 || hardLimit < softLimit) {
      throw new RangeError('invalid TTS text limits')
    }
    this.#softLimit = softLimit
    this.#hardLimit = hardLimit
  }

  push(text: unknown): readonly string[] {
    if (typeof text !== 'string') throw new TypeError('TTS text delta must be a string')
    const delta = [...text]
    if (delta.length > MAX_REALTIME_TEXT) throw new RangeError('TTS text delta is too large')
    this.#pending.push(...delta)
    const chunks: string[] = []
    while (this.#pending.length > 0) {
      const boundary = this.flushBoundary()
      if (boundary === null && this.#pending.length < this.#hardLimit) break
      const end = boundary === null ? this.#hardLimit : Math.min(boundary, this.#hardLimit)
      chunks.push(this.#pending.slice(0, end).join(''))
      this.#pending = this.#pending.slice(end)
      this.#first = false
    }
    return chunks
  }

  finish(): readonly string[] {
    if (this.#pending.length === 0) return []
    const pending = this.#pending.join('')
    this.#pending = []
    this.#first = false
    return [pending]
  }

  private flushBoundary(): number | null {
    for (let index = 0; index < this.#pending.length; index += 1) {
      if (!boundaries.has(this.#pending[index]!)) continue
      const end = index + 1
      if (this.#first || end >= this.#softLimit) return end
    }
    return null
  }
}

export type DoubaoTtsFailureCode =
  | 'configuration'
  | 'connect'
  | 'handshake'
  | 'session'
  | 'receive'
  | 'close'

export class DoubaoTtsFailure extends Error {
  readonly code: DoubaoTtsFailureCode

  constructor(code: DoubaoTtsFailureCode) {
    super(`Doubao TTS ${code} failure`)
    this.name = 'DoubaoTtsFailure'
    this.code = code
  }
}

export interface TtsAudio {
  readonly pcm: Uint8Array
}

export function ttsHeaders(input: {
  readonly apiKey: string
  readonly resourceId: string
  readonly idFactory?: () => string
}): Readonly<Record<string, string>> {
  const idFactory = input.idFactory ?? randomUUID
  if (!ttsNonblank(input.apiKey) || !ttsNonblank(input.resourceId)) {
    throw new DoubaoTtsFailure('configuration')
  }
  let connectId: string
  try {
    connectId = idFactory()
  } catch {
    throw new DoubaoTtsFailure('configuration')
  }
  if (!ttsNonblank(connectId) || !isWellFormed(connectId)) {
    throw new DoubaoTtsFailure('configuration')
  }
  return Object.freeze({
    'X-Api-Key': input.apiKey,
    'X-Api-Resource-Id': input.resourceId,
    'X-Api-Connect-Id': connectId,
  })
}

export interface DoubaoTtsClientOptions {
  readonly endpoint: string
  readonly apiKey: string
  readonly resourceId: string
  readonly voice: string
  readonly outputSampleRate?: 24_000
  readonly connectTimeoutMs?: number
  readonly receiveTimeoutMs?: number
  readonly connector?: VolcBinaryConnector
  readonly idFactory?: () => string
}

export class DoubaoTtsClient {
  readonly #options: Required<Omit<DoubaoTtsClientOptions, 'connector' | 'idFactory'>>
  readonly #connector: VolcBinaryConnector
  readonly #idFactory: () => string

  constructor(options: DoubaoTtsClientOptions) {
    const outputSampleRate = options.outputSampleRate ?? 24_000
    const connectTimeoutMs = options.connectTimeoutMs ?? DEFAULT_VOLC_CONNECT_TIMEOUT_MS
    const receiveTimeoutMs = options.receiveTimeoutMs ?? DEFAULT_VOLC_RECEIVE_TIMEOUT_MS
    if (!ttsNonblank(options.endpoint) || !ttsNonblank(options.apiKey)
      || !ttsNonblank(options.resourceId) || !ttsNonblank(options.voice)
      || outputSampleRate !== 24_000 || !ttsPositiveMilliseconds(connectTimeoutMs)
      || !ttsPositiveMilliseconds(receiveTimeoutMs)) {
      throw new DoubaoTtsFailure('configuration')
    }
    this.#options = {
      endpoint: options.endpoint,
      apiKey: options.apiKey,
      resourceId: options.resourceId,
      voice: options.voice,
      outputSampleRate,
      connectTimeoutMs,
      receiveTimeoutMs,
    }
    this.#connector = options.connector ?? webSocketVolcBinaryConnector
    this.#idFactory = options.idFactory ?? randomUUID
  }

  async open(signal?: AbortSignal): Promise<DoubaoTtsSession> {
    throwTtsIfAborted(signal)
    const connectionSignal = signal ?? new AbortController().signal
    let socket: VolcBinarySocket | undefined
    let phase: DoubaoTtsFailureCode = 'connect'
    try {
      socket = await this.#connector({
        endpoint: this.#options.endpoint,
        headers: {...ttsHeaders({
          apiKey: this.#options.apiKey,
          resourceId: this.#options.resourceId,
          idFactory: this.#idFactory,
        })},
        openTimeoutMs: this.#options.connectTimeoutMs,
        closeTimeoutMs: DEFAULT_VOLC_CLOSE_TIMEOUT_MS,
        maxFrameBytes: MAX_VOLCENGINE_WIRE_FRAME_BYTES,
        signal: connectionSignal,
      })
      phase = 'handshake'
      await socket.send(message(EventType.START_CONNECTION), signal)
      await expectTtsEvent(
        socket, EventType.CONNECTION_STARTED, null, this.#options.receiveTimeoutMs, signal,
      )
      const sessionId = ttsId(this.#idFactory)
      await socket.send(message(EventType.START_SESSION, sessionId, ttsPayload({
        voice: this.#options.voice,
        sampleRate: this.#options.outputSampleRate,
        event: EventType.START_SESSION,
        userId: ttsId(this.#idFactory),
      })), signal)
      await expectTtsEvent(
        socket, EventType.SESSION_STARTED, sessionId, this.#options.receiveTimeoutMs, signal,
      )
      return new DoubaoTtsSession({
        socket,
        sessionId,
        voice: this.#options.voice,
        sampleRate: this.#options.outputSampleRate,
        receiveTimeoutMs: this.#options.receiveTimeoutMs,
        idFactory: this.#idFactory,
      })
    } catch (error) {
      if (socket !== undefined) {
        try {
          await socket.close()
        } catch {
          // The original connection or handshake verdict remains authoritative.
        }
      }
      throwTtsIfAborted(signal)
      if (error instanceof DoubaoTtsFailure) throw error
      throw new DoubaoTtsFailure(phase)
    }
  }
}

export class DoubaoTtsSession {
  readonly #socket: VolcBinarySocket
  readonly #sessionId: string
  readonly #voice: string
  readonly #sampleRate: 24_000
  readonly #receiveTimeoutMs: number
  readonly #idFactory: () => string
  #writing: Promise<void> = Promise.resolve()
  #terminal: 'finish' | 'cancel' | undefined
  #closed = false
  #eventsClaimed = false
  #closePromise: Promise<void> | undefined

  constructor(input: {
    readonly socket: VolcBinarySocket
    readonly sessionId: string
    readonly voice: string
    readonly sampleRate: 24_000
    readonly receiveTimeoutMs: number
    readonly idFactory: () => string
  }) {
    this.#socket = input.socket
    this.#sessionId = input.sessionId
    this.#voice = input.voice
    this.#sampleRate = input.sampleRate
    this.#receiveTimeoutMs = input.receiveTimeoutMs
    this.#idFactory = input.idFactory
  }

  sendText(text: string, signal?: AbortSignal): Promise<void> {
    if (!ttsNonblank(text) || !isWellFormed(text) || [...text].length > MAX_REALTIME_TEXT) {
      return Promise.reject(new DoubaoTtsFailure('session'))
    }
    return this.#serialized(async () => {
      throwTtsIfAborted(signal)
      if (this.#terminal !== undefined || this.#closed) throw new DoubaoTtsFailure('session')
      try {
        await this.#socket.send(message(EventType.TASK_REQUEST, this.#sessionId, ttsPayload({
          voice: this.#voice,
          sampleRate: this.#sampleRate,
          event: EventType.TASK_REQUEST,
          userId: ttsId(this.#idFactory),
          text,
        })), signal)
      } catch {
        throwTtsIfAborted(signal)
        throw new DoubaoTtsFailure('session')
      }
    })
  }

  finish(signal?: AbortSignal): Promise<void> {
    return this.#terminalAction('finish', EventType.FINISH_SESSION, signal)
  }

  cancel(signal?: AbortSignal): Promise<void> {
    return this.#terminalAction('cancel', EventType.CANCEL_SESSION, signal)
  }

  async *events(signal?: AbortSignal): AsyncIterable<TtsAudio> {
    if (this.#eventsClaimed) throw new DoubaoTtsFailure('session')
    this.#eventsClaimed = true
    while (!this.#closed) {
      const raw = await ttsReceiveWithTimeout(
        this.#socket, this.#receiveTimeoutMs, signal, 'receive',
      )
      let decoded: VolcMessage
      try {
        decoded = VolcMessage.unmarshal(raw)
      } catch {
        throw new DoubaoTtsFailure('receive')
      }
      if (decoded.sessionId !== null && decoded.sessionId !== this.#sessionId) continue
      if (decoded.messageType === MessageType.AUDIO_ONLY_SERVER) {
        if (decoded.payload.byteLength === 0) continue
        try {
          yield {pcm: volcengineOutputPcm(decoded.payload).pcm}
        } catch {
          throw new DoubaoTtsFailure('receive')
        }
        continue
      }
      if (decoded.event === EventType.SESSION_FINISHED) return
      if (decoded.messageType === MessageType.ERROR
        || decoded.event === EventType.SESSION_FAILED
        || decoded.event === EventType.CONNECTION_FAILED) {
        throw new DoubaoTtsFailure('receive')
      }
    }
  }

  close(): Promise<void> {
    if (this.#closePromise !== undefined) return this.#closePromise
    this.#closed = true
    this.#closePromise = this.#performClose()
    return this.#closePromise
  }

  #terminalAction(
    terminal: 'finish' | 'cancel',
    event: typeof EventType.FINISH_SESSION | typeof EventType.CANCEL_SESSION,
    signal?: AbortSignal,
  ): Promise<void> {
    return this.#serialized(async () => {
      throwTtsIfAborted(signal)
      if (this.#closed) throw new DoubaoTtsFailure('session')
      if (this.#terminal !== undefined) return
      this.#terminal = terminal
      try {
        await this.#socket.send(message(event, this.#sessionId), signal)
      } catch {
        throwTtsIfAborted(signal)
        throw new DoubaoTtsFailure('session')
      }
    })
  }

  async #performClose(): Promise<void> {
    let failed = false
    try {
      await this.#serialized(async () => {
        try {
          await this.#socket.send(message(EventType.FINISH_CONNECTION))
        } catch {
          failed = true
        }
      })
    } finally {
      try {
        await this.#socket.close()
      } catch {
        failed = true
      }
    }
    if (failed) throw new DoubaoTtsFailure('close')
  }

  #serialized(operation: () => Promise<void>): Promise<void> {
    const result = this.#writing.then(operation)
    this.#writing = result.then(() => undefined, () => undefined)
    return result
  }
}

async function expectTtsEvent(
  socket: VolcBinarySocket,
  expected: EventType,
  sessionId: string | null,
  timeoutMs: number,
  signal?: AbortSignal,
): Promise<void> {
  const raw = await ttsReceiveWithTimeout(socket, timeoutMs, signal, 'handshake')
  let decoded: VolcMessage
  try {
    decoded = VolcMessage.unmarshal(raw)
  } catch {
    throw new DoubaoTtsFailure('handshake')
  }
  if (decoded.messageType !== MessageType.FULL_SERVER_RESPONSE || decoded.event !== expected
    || sessionId !== null && decoded.sessionId !== sessionId) {
    throw new DoubaoTtsFailure('handshake')
  }
}

async function ttsReceiveWithTimeout(
  socket: VolcBinarySocket,
  timeoutMs: number,
  signal: AbortSignal | undefined,
  code: 'handshake' | 'receive',
): Promise<Uint8Array> {
  throwTtsIfAborted(signal)
  const timeout = new AbortController()
  const combined = signal === undefined ? timeout.signal : AbortSignal.any([timeout.signal, signal])
  const timer = setTimeout(() => timeout.abort(), timeoutMs)
  try {
    return await socket.receive(combined)
  } catch {
    throwTtsIfAborted(signal)
    throw new DoubaoTtsFailure(code)
  } finally {
    clearTimeout(timer)
  }
}

function throwTtsIfAborted(signal: AbortSignal | undefined): void {
  if (signal?.aborted !== true) return
  if (signal.reason instanceof Error) throw signal.reason
  throw new DOMException('This operation was aborted', 'AbortError')
}

function ttsPositiveMilliseconds(value: number): boolean {
  return Number.isSafeInteger(value) && value > 0
}

function ttsNonblank(value: unknown): value is string {
  return typeof value === 'string' && stripLikePython(value).length > 0
}

function ttsId(idFactory: () => string): string {
  let value: string
  try {
    value = idFactory()
  } catch {
    throw new DoubaoTtsFailure('configuration')
  }
  if (!ttsNonblank(value) || !isWellFormed(value)) throw new DoubaoTtsFailure('configuration')
  return value
}

function message(event: EventType, sessionId?: string, payload?: object): Uint8Array {
  return new VolcMessage({
    messageType: MessageType.FULL_CLIENT_REQUEST,
    event,
    payload: new TextEncoder().encode(JSON.stringify(payload ?? {})),
    ...(sessionId === undefined ? {} : {sessionId}),
  }).marshal()
}

function ttsPayload(input: {
  readonly voice: string
  readonly sampleRate: 24_000
  readonly event: EventType
  readonly userId: string
  readonly text?: string
}): object {
  return {
    event: input.event,
    user: {uid: input.userId},
    namespace: 'BidirectionalTTS',
    req_params: {
      speaker: input.voice,
      audio_params: {
        format: 'pcm', sample_rate: input.sampleRate, enable_timestamp: false,
      },
      additions: JSON.stringify({disable_markdown_filter: false}),
      ...(input.text === undefined ? {} : {text: input.text}),
    },
  }
}
